ksqldb-common/src/main/java/io/confluent/ksql/name/ColumnNames.java

```diff
@@ -41,11 +50,24 @@ public static ColumnName aggregateColumn(final int idx) {
   }
 
   /**
-   * Where the user hasn't specified an alias for an expression in a SELECT we generate them using
-   * this method. This value is exposed to the user in the output schema
+   * Create a generator that will build column aliases in the form {@code KSQL_COL_x}.
+   *
+   * <p>Names are guaranteed not to clash with any existing columns in the {@code sourceSchemas}.
+   *
+   * <p>Used where the user hasn't specified an alias for an expression in a SELECT. This generated
+   * column names are exposed to the user in the output schema.
+   *
+   * @param sourceSchemas the stream of source schemas.
+   * @return a generator of unique column names.
    */
-  public static ColumnName generatedColumnAlias(final int idx) {
-    return ColumnName.of(GENERATED_ALIAS_PREFIX + idx);
+  public static Supplier<ColumnName> columnAliasGenerator(
+      final Stream<LogicalSchema> sourceSchemas
+  ) {
+    final Set<Integer> used = generatedAliasIndexes(sourceSchemas)
+        .boxed()
+        .collect(Collectors.toSet());
+
+    return new AliasGenerator(0, used)::next;
```

Comment:
nit: if we pass 1 here we should get names more consistent with what we had before (start at offset 1), and a smaller test diff.

Reply:
I thought that too and tried it. However, the old code started at zero too.  If we switch this to 1 we just get a different set of tests to change.

The changes are because the name is no longer controlled by the index of the select expresssion.